Refactor definitions of ADTs in rustc::middle::def

petrochenkov opened the pull request:
All structs and their constructors are defined as `DefStruct`.
`DefTy` is splitted into `DefEnum` and `DefTyAlias`.
Ad hoc flag `bool is_structure` is removed from `DefVariant`, it was required in one place in resolve and could be obtained by other means.
Flag `bool is_ctor` is removed from `DefFn`, it wasn't really used for constructors outside of metadata decoding. 

Observable effects:
More specific error messages are selected in some cases.
Two name resolution bugs fixed (https://github.com/rust-lang/rust/issues/30992 and FIXME in compile-fail/empty-struct-braces-expr.rs).

Fixes https://github.com/rust-lang/rust/issues/30992
Closes https://github.com/rust-lang/rust/issues/30361

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

nagisa commented on src/librustc/middle/def.rs:
```diff
@@ -29,8 +29,9 @@ pub enum Def {
              ast::NodeId), // node id of variable
-    DefVariant(DefId /* enum */, DefId /* variant */, bool /* is_structure */),
-    DefTy(DefId, bool /* is_enum */),
+    DefVariant(DefId /* enum */, DefId /* variant */),
```
Since you’re refactoring here, might it also be a good time to move from tuple-variants to struct-variants, so comments like these aren’t necessary?

petrochenkov replied:
Yeah, and rename `DefVariant` to `Def::Variant` etc.
It would be a huge code churn and I'd like to separate semantic changes done in this PR from pure renaming. But I can do it in a separate commits if reviewer agrees.

arielb1 commented:
Renaming `DefVariant` to `Def::Variant` etc. would be fine.

Also, test needed.

r+ modulo test.

petrochenkov commented:
Updated with a test and renamed variants.

@nagisa 
Making `Def` variants struct-like is too much manual work, unfortunately. I've omitted this part.

nagisa commented:
@bors r=arielb 2084c2c

nagisa commented:
@bors r=arielb1 2084c2c

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 2084c2c has been approved by `arielb1`

<!-- @bors r=arielb1 2084c2c33a47407c5ec514898d72701cf0233ba9 -->

bors commented:
:hourglass: Testing commit 2084c2c with merge 51108b6...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7767), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7787), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1036), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7726), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7741), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7727), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/944), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2895), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7789), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7785), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7749), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2775), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2790), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2786), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2765), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2003), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2768)
